book/src/spec.md

```diff
@@ -0,0 +1,628 @@
+asserted.
+
+### Signal omission
+
+Not all signals are always needed. The following table shows the conditions for
+a signal to be relevant. If the condition is false, the signal may be omitted
+on the interface. When two interfaces with differing but compatible
+complexities are connected together, the default value specified in the table
+below must be driven for any omitted signals.
+
+| Name    | Condition                                   | Default   |
+|---------|---------------------------------------------|-----------|
+| `valid` | \\(\top\\)                                  | `'1'`     |
+| `ready` | \\(\top\\)                                  | `'1'`     |
+| `data`  | \\(E > 0\\)                                 | all `'0'` |
+| `last`  | \\(D \ge 1\\)                               | all `'1'` |
+| `stai`  | \\(C \ge 6 \wedge N > 1\\)                  | 0         |
+| `endi`  | \\((C \ge 5 \vee D \ge 1) \wedge N > 1\\)   | \\(N-1\\) |
+| `strb`  | \\(C \ge 7 \vee D \ge 1\\)                  | all `'1'` |
+| `user`  | \\(U > 0\\)                                 | all `'0'` |
+
+### Interface conventions
+
+When there are name conflicts (for instance due to multiple physical streams
+existing on the hierarchical level of an interface) the signal names must be
+prefixed by a unique stream name, delimited using an underscore.
+
+The canonical representation for the `data` and `user` signals is the LSB-first
+concatenation of the contained fields. For \\(N > 1\\), the lanes are
+concatenated LSB first, such that the lane index is major and the field index
+is minor.
+
+Where applicable, the signals must be listed in the following order: `dn`,
+`up`, `valid`, `ready`, `data`, `last`, `stai`, `endi`, `strb`, `user`.
+
+> This is mostly just a consistency thing, primarily because it helps to get
+> used to a single ordering when interpreting simulation waveforms. It may also
+> simplify tooling.
+
+The signal names must be lowercase.
+
+> This is for interoperability between languages that differ in case
+> sensitivity.
+
+#### Alternative representation
+
+To improve code readability in hardware definition languages supporting array
+and aggregate constructs (record, struct, ...), the following changes are
+permissible. However, it is recommended to fall back to the conventions above
+for interoperability with other streamlets on the "outer" interfaces of an IP
+block.
+
+ - `valid`, `data`, `last`, `stai`, `endi`, `strb`, and `user` may be bundled
+   in an aggregate type named `<stream-name>_dn_type` with signal name
+   `<stream-name>_dn`.
+
+ - `ready` may be "bundled" in an aggregate type named `<stream-name>_up_type`
+   with signal name `<stream-name>_up` for symmetry.
+
+ - If the language allows for signal direction reversal within a bundle, all
+   stream signals may also be bundled into a single type named
+   `<stream-name>_type`, with `ready` in the reverse direction.
```

Comment:
Perhaps we can do a little verilog modport example

Reply:
>It's clear enough to me but we could also leave out any signals that are never omitted from the table.

-> https://github.com/abs-tudelft/tydi/pull/8#discussion_r373407786

>Perhaps we can do a little verilog modport example

I don't even want to know what this means, but feel free to PR something into this branch.